Miscellaneous/The-Prestige/TusharJoshi.cpp

```diff
@@ -0,0 +1,175 @@
+#include <bits/stdc++.h>
+#define ll long long
+
+using namespace std;
+
+const int N = 2e5 + 5;
+
+struct node {
+    int left, right, par, n;
+    int rev;
+    ll fa, fb, va, vb;
+    node()
+        : left(0), right(0), rev(0), fa(0), fb(0), par(0), n(0), va(0), vb(0) {}
+};
+
+node nodes[N];
+
+void g(int u) {
+    nodes[u].fa =
+        nodes[nodes[u].left].fa + nodes[nodes[u].right].fa + nodes[u].va;
+    nodes[u].fb =
+        nodes[nodes[u].left].fb + nodes[nodes[u].right].fb + nodes[u].vb;
+    nodes[u].n = nodes[nodes[u].left].n + nodes[nodes[u].right].n + 1;
+}
+
+void splay(int u) {
+    int p = nodes[u].par;
+    int pp = nodes[p].par;
+
+    if (nodes[p].right == u) {
+        nodes[p].right = nodes[u].left;
+        nodes[nodes[u].left].par = p;
+        nodes[u].left = p;
+    } else {
+        nodes[p].left = nodes[u].right;
+        nodes[nodes[u].right].par = p;
+        nodes[u].right = p;
+    }
+
+    if (nodes[pp].right == p) {
+        nodes[pp].right = u;
+    } else {
+        nodes[pp].left = u;
+    }
+    nodes[u].par = pp;
+    nodes[p].par = u;
+
+    g(p);
+    g(u);
+}
+
+void flip(int u) {
+    swap(nodes[u].left, nodes[u].right);
+    swap(nodes[u].va, nodes[u].vb);
+    swap(nodes[u].fa, nodes[u].fb);
+    nodes[u].rev ^= 1;
+}
+
+void push(int v) {
+    if (nodes[v].rev) {
+        flip(nodes[v].left);
+        flip(nodes[v].right);
+        nodes[v].rev = 0;
+    }
+}
+
+int find(int root, int l) {
```

Comment:
It would be better if Splay are added into the template first and using that this question is solved as it is highly unlikely to write splay from scratch in live contests

Reply:
Incase you feel we should merge this solution right now and may be add to the template later, would also work